docs/designs/topology.md

```diff
@@ -0,0 +1,916 @@
+    2. **Operator Configuration**: Selects active topology via `--topology-domain-name` argument
+    3. **TopologyConstraint**: User-specified packing requirements in workloads (PodCliqueSet, PodCliqueScalingGroup,
+       PodClique)
+
+**Key Feature**: Grove provides automatic out-of-box topology optimization by generating preferred packing constraints
+at all levels, even without user configuration. Users can optionally specify required constraints for strict placement
+requirements.
+
+## Goals
+
+- Provide flexible, cluster-agnostic topology hierarchy definition via TopologyDomain CRD
+    - Enable packing constraints for network locality across all Grove scalable resources
+    - Support multiple topology configurations for different environments
+    - Automatic Kueue Topology generation for KAI scheduler integration
+    - Immutable topology configuration ensuring scheduling consistency
+    - Hierarchical constraint validation (child stricter than parent)
+
+## Non-Goals
+
+- Spread constraints across topology domains (ReplicaSpreadDomain)
+    - Root domain constraints for entire resource (RootDomain)
+    - Ratio-based affinity groups between scaling groups (AffinityGroups with PackRatio)
+    - Dynamic topology reconfiguration after creation
+    - Per-workload topology domain selection
+    - Automatic topology inference from workload characteristics
+
+## Proposal
+
+### High-Level Approach
+
+Grove implements topology-aware scheduling through three key components:
+
+1. **TopologyDomain CRD**: Cluster-scoped resource defining topology hierarchy
+    - Admin creates TopologyDomain with ordered list of topology levels
+    - Each level maps friendly name (e.g., "rack") to node label key (e.g., "topology.kubernetes.io/rack")
+    - Multiple TopologyDomains supported for different environments
+
+    2. **Operator Configuration**: References TopologyDomain by name
+        - Operator argument `--topology-domain-name=default` selects which TopologyDomain to use
+        - All workload validation performed against configured TopologyDomain
+        - Enables switching between topologies without changing workloads
+
+    3. **Workload API (TopologyConstraint)**: Users specify packing requirements
+        - PodCliqueSet, PodCliqueScalingGroup, and PodClique each have TopologyConstraint field
+        - Users reference level names from TopologyDomain (e.g., `packDomain: "rack"`)
+        - No direct TopologyDomain reference needed in workloads
+
+### Component Interactions
+
+```
+TopologyDomain CRD ──┐
+  (admin creates)    │
+                     │
+Operator Config ─────┼──> Operator validates PackDomain
+  (--topology-       │    against TopologyDomain.Spec.Levels
+   domain-name)      │
+                     │
+PodCliqueSet ────────┘
+  (packDomain: "rack")
+```
+
+### Automatic Optimization
+
+**Out-of-Box Optimization:**
+
+- Operator automatically generates **preferred** constraints using strictest topology level (e.g., "host")
```

Comment:
If the user is provided an explicit API to define network pack constraints then shouldn't it be wise to let the user take the call w.r.t pack constraints? Why assume `host`? When you mention `Out of the box optimization` then assuming `host` might be completely wrong as we are not looking at the resource requirements upfront and the feasibility to schedule such a workload given the available capacity and where the capacity is available. 

We are trying to do in Grove what the scheduler should ideally do. On top of that assuming Host for a workload that will never fit in a single Node cannot be termed as `out of the box optimization`.